### DISCUSSION TITLE
Game not checks autofill tags on crates within crate shelves

### DISCUSSION THREAD

**Discussion by Xardion495** [FakeFishGames/Barotrauma#13683] [Bug reports]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [ ] My issue happened while using mods.

### What happened?

If crates placed inside the shelves, the game not checks autofill tags on them.
Please change this behavior.

### Reproduction steps

-

### Bug prevalence

Happens every time I play

### Single player or multiplayer?

Other

### -

Submarine spawner

### Version

v1.2.7.0 (Winter Update hotfix)

### Which operating system did you encounter this bug on?

Windows

**Comment by Regalis11** [comment]
Thank you for the report!

I'm not entirely sure I understand the issue. I tested adding the tag "engcab" to a metal crate and placing it in a crate shelf, and it seems to get filled with engineering supplies as expected. Do you mean something else by "autofill tags"?

**Comment by Xardion495** [comment]
Filling work as expected but the game still spawn a support crates at the beginning of campaign which must not be happens if autofill is enabled on corresponding containers.

For example.
- If we add 'engcab' tag to locker and enable autofill - corresponding support crate will not be spawn. It's ok.
- If we add 'engcab' tag to crate, enable autofill and leave it on the floor - corresponding support crate will not be spawn. It's ok.
- If we add 'engcab' tag to crate, enable autofill and place it within the shelf - corresponding support crate will be spawn. It's not ok.

**Comment by Regalis11** [comment]
What do you mean by a support crate? The extra crates that spawn if there's not enough space for the supplies in cabinets or other containers?

To me it doesn't seem like there's necessarily a bug here. If there's not enough space for the initial supplies in that crate tagged as "engcab", the rest of the items should spawn in a new crate.

**Comment by Xardion495** [comment]
Well, the truth is even there is enough space in crates the game will spawns extra ones if crates placed in shelves.
I'll show in the pictures.
1. Cargo spawnpoint where initial crates spawns
![20240103173955_1](https://github.com/Regalis11/Barotrauma/assets/17722887/b090730d-65d9-4ce7-97bd-d3867601bf85)
2. For example add manually two crates. One with 'secarmcab' tag and another one with 'toxcab'+'toxcontainer' tags. Both are empty. Both have an autofill properties enabled. Leave them just on the floor.
![20240103170513_1](https://github.com/Regalis11/Barotrauma/assets/17722887/1fbf3b46-363a-4cb6-8fcb-aacb2fbf49a3)
![20240103170521_1](https://github.com/Regalis11/Barotrauma/assets/17722887/f862adff-f091-46b5-bf61-80822ea6643f)
![20240103174019_1](https://github.com/Regalis11/Barotrauma/assets/17722887/0171cd5b-32ef-420f-bd72-86b566526f32)
![20240103170530_1](https://github.com/Regalis11/Barotrauma/assets/17722887/5856aff4-2d4d-4548-9adb-5dd0f6ab2ab6)
3. Press 'test' button and check cargo hull. We will see that no initial crates spawned.
![20240103170557_1](https://github.com/Regalis11/Barotrauma/assets/17722887/3fd92972-ccce-42eb-8703-eca55b4f4894)
4. Check manually added crates. Both are filled corresponding to their tags.
![20240103174212_1](https://github.com/Regalis11/Barotrauma/assets/17722887/fb79f46a-aa60-4717-8f79-daa0bd2e4c43)
![20240103174216_1](https://github.com/Regalis11/Barotrauma/assets/17722887/b0a5ce5d-4db4-4d69-a239-12857e54594e)
5. Now make another test. Place crates within the shelf and launch the test again.
![20240103180930_1](https://github.com/Regalis11/Barotrauma/assets/17722887/d1b64dfa-ef7a-4d23-922c-460e37be8452)
6. Check crates within the shelf. Both filled as expected.
![20240103181059_1](https://github.com/Regalis11/Barotrauma/assets/17722887/7dddfaf1-0039-4d65-b212-81347eb4ad17)
7. Check cargo hull. We will see that initial crates has been spawned, despite the fact that there are such "initial" crates on the boat
![20240103181110_1](https://github.com/Regalis11/Barotrauma/assets/17722887/362d57ea-754f-408a-aa5d-d9b9941ef11d)
![20240103181118_1](https://github.com/Regalis11/Barotrauma/assets/17722887/24e142f8-a9ec-4183-a2f9-b8e3e8b431cd)

Thus we can conclude that the game when creating initial crates considering the tags on lockers and crates but not considering tags on crates if they within the shelves.

**Comment by Regalis11** [comment]
In those screenshots, the weapon crate seems to be already full, so the game needs to spawn an extra one to have enough space for all the supplies. In the last pic, you're holding a medical crate, not a chemical crate - again, the game needs to spawn a medical crate because there's no other place to put the medical items in.

There does seem to be something off here though, I'm not sure why the extra crates aren't spawning when you leave the manually placed crates on the floor.

**Comment by Regalis11** [comment]
Hmm, I unfortunately still couldn't reproduce this. :/

I'm getting the exact same number of crates, and the contents are also identical, regardless if the manually placed crates are in a crate shelf or not. So I suspect this requires some more specific conditions to occur.
![image](https://github.com/Regalis11/Barotrauma/assets/14351070/85cea791-cf97-4bb1-8bba-7cc4760d0458)

**Comment by Xardion495** [comment]
To see the difference just add 'medfabcab' tag to one of the custom crates.
![20240106031110_1](https://github.com/Regalis11/Barotrauma/assets/17722887/5f93d2b4-7e16-41d0-89bd-c32e7464134a)
![20240106114744_1](https://github.com/Regalis11/Barotrauma/assets/17722887/9da8e41f-8b1f-43c1-bb62-04997c1b5fb5)

I checked a code block responsible for initial crates. The bottom line is that the game does not spawn the initial crates exactly, but looks for suitable containers for "start item set" spawning. And if there is no such one or there is not enough space in container, it spawns a new container (crate) and place the item there.

Default start item set is "normal" ([Barotrauma/BarotraumaShared/SharedSource/GameSession
/AutoItemPlacer.cs:70](https://github.com/Regalis11/Barotrauma/blob/8ea2b47889ede027c2214ce95f4f0d26b67ee450/Barotrauma/BarotraumaShared/SharedSource/GameSession/AutoItemPlacer.cs#L70C22-L70C22))
Set composition can be viewed in the file of the installed distribution ("**...\SteamLibrary\steamapps\common\Barotrauma\Content\Items\StartItems.xml**")
When the new item list is formed ([Barotrauma/BarotraumaShared/SharedSource/GameSession
/AutoItemPlacer.cs:104-123](https://github.com/Regalis11/Barotrauma/blob/8ea2b47889ede027c2214ce95f4f0d26b67ee450/Barotrauma/BarotraumaShared/SharedSource/GameSession/AutoItemPlacer.cs#L104)) the next code block starts searching for the container for each item from that list ([Barotrauma/BarotraumaShared/SharedSource/GameSession/AutoItemPlacer.cs:124-141](https://github.com/Regalis11/Barotrauma/blob/8ea2b47889ede027c2214ce95f4f0d26b67ee450/Barotrauma/BarotraumaShared/SharedSource/GameSession/AutoItemPlacer.cs#L124))

Possible solution may be to modificate **FindContainerFor()** method in [Barotrauma/BarotraumaShared/SharedSource/Map/Submarine.cs:2120](https://github.com/Regalis11/Barotrauma/blob/8ea2b47889ede027c2214ce95f4f0d26b67ee450/Barotrauma/BarotraumaShared/SharedSource/Map/Submarine.cs#L2120) to add collection parsing for items with 'crateshelf' identifier or 'cargocontainer' tag
